packages/loopring_v3/contracts/impl/libexchange/ExchangeWithdrawals.sol

```diff
@@ -237,226 +231,78 @@ library ExchangeWithdrawals
+        for (uint i = 0; i < withdrawals.length; i += 7) {
+            uint data = uint(withdrawals.bytesToUint56(i));
 
-        // Only allow withdrawing on finalized blocks
-        require(blockIdx < S.numBlocksFinalized, "BLOCK_NOT_FINALIZED");
-        // Check if the withdrawals were already completely distributed
-        require(withdrawBlock.numWithdrawalsDistributed < withdrawBlock.blockSize, "WITHDRAWALS_ALREADY_DISTRIBUTED");
-
-        // Only allow the operator to distribute withdrawals at first, if he doesn't do it in time
-        // anyone can do it and get paid a part of the exchange stake
-        bool bOnlyOperator = now < withdrawBlock.timestamp + ExchangeData.MAX_TIME_TO_DISTRIBUTE_WITHDRAWALS();
-        if (bOnlyOperator) {
-            require(msg.sender == S.operator, "UNAUTHORIZED");
-        }
-
-        // Calculate the range of withdrawals we'll do
-        uint start = withdrawBlock.numWithdrawalsDistributed;
-        uint end = start.add(maxNumWithdrawals);
-        if (end > withdrawBlock.blockSize) {
-            end = withdrawBlock.blockSize;
-        }
+            // Extract the withdrawal data
+            uint16 tokenID = uint16((data >> 48) & 0xFF);
+            uint24 accountID = uint24((data >> 24) & 0xFFFFFF);
+            uint amount = (data & 0xFFFFFF).decodeFloat(24);
 
-        // Do the withdrawals
-        uint gasLimit = ExchangeData.MIN_GAS_TO_DISTRIBUTE_WITHDRAWALS();
-        uint totalNumWithdrawn = start;
-        while (totalNumWithdrawn < end && gasleft() >= gasLimit) {
-            // Don't check the return value here, the withdrawal is allowed to fail.
-            // The automatic token disribution by the operator is a best effort only.
-            // The account owner can always manually withdraw without any limits.
-            withdrawFromApprovedWithdrawal(
+            // Transfer the tokens
+            bool success = transferTokens(
                 S,
-                blockIdx,
-                withdrawBlock,
-                totalNumWithdrawn,
+                accountID,
+                tokenID,
+                amount,
                 true
             );
-            totalNumWithdrawn++;
-        }
-        withdrawBlock.numWithdrawalsDistributed = uint16(totalNumWithdrawn);
-
-        // Fine the exchange if the withdrawals are done too late
-        if (!bOnlyOperator) {
-            // We use the stake of the exchange to punish withdrawals that are distributed too late
-            uint numWithdrawn = totalNumWithdrawn.sub(start);
-            uint totalFine = S.loopring.withdrawalFineLRC().mul(numWithdrawn);
-            // Burn 50% of the fine, reward the distributer the rest
-            uint amountToBurn = totalFine / 2;
-            uint amountToDistributer = totalFine - amountToBurn;
-            S.loopring.burnExchangeStake(S.id, amountToBurn);
-            S.loopring.withdrawExchangeStake(S.id, msg.sender, amountToDistributer);
+            if (!success) {
```

Comment:
we may want to emit an event if any withdrawal distribution failed in a block.

Reply:
The events (`WithdrawalCompleted` and `WithdrawalFailed`) are emitted in `ExchangeWithdrawals.transferTokens` so it is shared for other ways withdrawals are possible.